Language/Variables/Variable Scope & Qualifiers/volatile.adoc

```diff
@@ -55,6 +55,7 @@ There are several ways to do this:
 === Example Code
 // Describe what the example code is all about and add relevant code   ►►►►► THIS SECTION IS MANDATORY ◄◄◄◄◄
 
+Use a `volatile byte` to store the pin state in a single byte:
```

Comment:
I don't think this is a good description of the code.

In my experience "a `volatile byte`" is not really a term that is used. It would be more common to start the sentence "Use the `volatile` keyword".

Second, the purpose of this code is not to "store the pin state in a single byte". That would happen with or without the use of the `volitile` keyword. The purpose is to ensure that the compiler doesn't optimize out what it will consider to be the unused `state` variable.

Reply:
I do not understand the second part. Why would the compiler think the `state` variable is unused? Isn’t the reason of using `volatile` here that changes to the variable could be missed? In fact, using `volatile` in this code example should not make any (functional) difference because `loop()` only writes the pin after reading the variable, and every change to the variable will cause a pin change in the next loop.

`volatile` would, however, make a difference in the following case:

```c

// The LED only shows that the pin changed if the state variable is declared as volatile.
// Otherwise, updates to the variable are not seen inside the loo() funciton.
#define USE_VOLATILE

#define pin LED_BUILTIN

#ifdef USE_VOLATILE
volatile byte state = LOW;
#else
byte state = LOW;
#endif

void setup() {
  pinMode(pin, OUTPUT);
  attachInterrupt(digitalPinToInterrupt(2), toggle, CHANGE);
}

void loop() {
  bool changedInTheMeantime = false;
  byte originalState = state;
  delay(1000);
  if (originalState != state) {
    changedInTheMeantime = true;
  }
  
  digitalWrite(pin, changedInTheMeantime ? HIGH : LOW);
  
}

void toggle() {
  state = !state;
}
```

Reply:
>Isn’t the reason of using volatile here that changes to the variable could be missed?

Yes, that's what I mean.

>In fact, using volatile in this code example should not make any (functional) difference because loop() only writes the pin after reading the variable, and every change to the variable will cause a pin change in the next loop.

I'm no expert on the mysterious ways of compilers, but my understanding is that since `blink()` is only called from the ISR, the compiler might think that the value of `state` is never changed after its initialization to `LOW`, resulting in a loop like this:
```c++
void loop() {
  digitalWrite(pin, LOW);
}
```

Reply:
I quickly tested it with an Arduino, both ways work!

Changed the code sample – better this way?